### ISSUE TITLE
Evaluate U-shape Transformer Model For Image Enhancement

### ISSUE THREAD

**Issue by keenanjohnson** [Seattle-Aquarium/CCR_image_processing#15]
Previously, we have been only evaluating the U Net architecture for training auto encoder models (https://en.wikipedia.org/wiki/U-Net)

However, there are several other promising papers like the U Shape Transformer Model here: https://github.com/LintaoPeng/U-shape_Transformer_for_Underwater_Image_Enhancement

We should evaluate this and try to compare them.

**Comment by keenanjohnson** [comment]
This was processed on my laptop using the UIE GUI, which is a mac, but has a normal amount of RAM (16GB), so I'm confused why you all still see artifacts as I don't see them, which implies that either you aren't using the most update to date GUI or thee is a difference between windows and mac in the UIE gui. 

I don't believe @reidtosa has published any new versions of the datasets with a partitioned section.

**Comment by m-h-williams** [comment]
@keenanjohnson I’ve added a new SOP and supporting Python scripts to this repo to streamline our workflow for evaluating UIE-processed imagery against hand-edited training images.

This update includes:

- A full SOP ([linked here](https://github.com/Seattle-Aquarium/CCR_image_processing/blob/main/documents/comparison_UIE.md)) outlining the step-by-step process for preparing datasets, updating annotation paths, matching dataset splits, training YOLO classification models, and comparing performance metrics.

- A [new code/ folder](https://github.com/Seattle-Aquarium/CCR_image_processing/tree/main/code) containing three scripts detailed in the SOP:
   - `update_annotation_paths.py` — updates annotation JSON files to point to UIE image directories
   - `match_dataset_split.py` — ensures UIE and hand-edited datasets have identical train/val/test splits
   - `ML_metrics_comparison.py` — generates a clean comparison table of model performance metrics

These should make the workflow much smoother, more reproducible, and quicker for anyone on the team to run. Let me know if you have ideas, questions, or anything you'd like to tweak!

**Comment by reidtosa** [comment]
@keenanjohnson, overall, I've been very impressed by the performance our U-Shape Transformer models; however, I've also picked up on a consistent trend in the outputs. The same several images always develop "smudges"/blurry spots in the same location regardless of epoch or which trained U-Shape model we apply. Shown below are blurry spots in 2024_10_08_11-40-27 after applying the "all01" model at epochs 25 and 158 (trained on combined flat and complex image sets) and the "flat06" best model (trained on only the flat image set).

<p align="center">
<img width=60% alt="Image" src="https://github.com/user-attachments/assets/fed62337-fca9-4613-90b9-65539132332b" />
</p>
<img width=33% alt="Image" alt="Image" src="https://github.com/user-attachments/assets/e7b1fda2-dc17-4cb0-898a-d7488d30236b" />
<img width=33% alt="Image" alt="Image" src="https://github.com/user-attachments/assets/6ac28cce-f771-4144-aa24-32ea46f748a1" />
<img width=33% alt="Image" alt="Image" src="https://github.com/user-attachments/assets/e891f6d9-27ec-4411-8918-21db240edc5f" />
<img width=33% alt="Image" src="https://github.com/user-attachments/assets/9753a977-4b12-41c8-a397-98ddb5a4ce3f" />
<img width=33% alt="Image"  src="https://github.com/user-attachments/assets/cbaa573e-52d3-4236-a22a-c2f07f4aae51" />
<img width=33% alt="Image" alt="Image" src="https://github.com/user-attachments/assets/2400f5f3-133a-4cbe-8168-a3cdc5aed7e8" />

It does seem like these blurry areas worsen over time, but from what I can tell they start to occur early in development. Furthermore, many of them occur in the center portion of the image, which should retain the most detail.

Looking back, Keenan, we also noticed these spots in some of the initial testing images you shared for U-Shape.
> 2. I trained a model using this network using 1024 patches. This network trained for 150 iterations. The time on a Mac studio with M3 processor and 96GB of shared memory was around 1 hour per iteration, so this took roughly 150 hours. I have posted this model to our hugging face account here: https://huggingface.co/Seattle-Aquarium/ushape_transformer_1024_150_iterations
> 3. I processed the usual test images with this model and upload the results to this folder: https://drive.google.com/drive/folders/1eiF4mKTlcDup2T8ikP7lDjWMblt4-1Zv?usp=drive_link
<p align="center">
<img width="478" height="266" alt="Image" src="https://github.com/user-attachments/assets/bc10d7a9-5401-464f-baf4-aa6625eb2661" />
</p>

Interestingly, a little over half of the images have no blur at all and look great. Any thoughts on whether these artifacts can be worked out somewhere down the line? Any clues as to why this happens only to a selection of the images? I'll upload the models above to Hugging Face in case you'd like to see for yourself.

**Comment by keenanjohnson** [comment]
Hmmmm that is quite interesting @reidtosa . I have to assume that this artifact is an output of the model itself, rather than the GUI or post processing where we stitch patches together since we don't see it in the other model type correct?

Why that happens is unclear to me. One theory is that there is something about those spots which is tricking the model into processing in some other way that is incorrect (perhaps it estimates depth incorrectly, etc), but that's really just conjecture I would say.

**Comment by keenanjohnson** [comment]
Hey all! I've had a few moments here and there to work on this further and I have some learnings to share and hoping to have more time to come to an actual conclusion. I shared some of this with Zach last week, but finally had a moment to document. 

The main question we've all had is whether these blurry artifacts are some sort of bug in the blending algorithm that processes tiles when using the GUI or some inherent issue with the trained model. To help get to the bottom of this, I first edited a version of the GUI script to do no overlapping or blending of the tiles during processing. It just tiles up the image and processes each through the trained model. The results as an example are shown in this image where you can clearly see each tile (this also illustrates why the blending / overlap is required as the color differences tile to tile can very widely).

<img width="882" height="782" alt="Image" src="https://github.com/user-attachments/assets/be4a0cdb-195e-40b5-8b4a-b0bc0e817c74" />

Now, if the blurriness issue was purely from the blending or overlap process the GUI does, we would expect to see all non blurry tiles in this zero overlap method. But as you can see, in the problem spots the whole tile is clearly blurry. The processing python script does no modification to the pixel colors coming out of the model, so I think this conclusively proves that the blurry patches are an artifact of the trained U-shape transformer model in some way. 

<img width="794" height="619" alt="Image" src="https://github.com/user-attachments/assets/6487abe7-f9d5-46f4-a3c5-a2a9b39521ef" />

Now what to do about that haha. I think there are a few possibilities:

1. The blurriness is because the model does not have sufficient context as to the surrounding area and that by increasing the size of image patches we train the model on, we could give the model enough information to fix these blurry spots. 
2. This is an inherent flaw of the current u-shape transformer model and we need to go back to the more basic U-net or try implementing another model from one of the papers.

Option 1 above is easy enough to test, and I think the next step is to try and increase the patch size for training, which would either confirm or rule out number 1, leaving only option 2.

**Comment by zhrandell** [comment]
Thanks for this, @keenanjohnson! Super interesting to see the base patches prior to the blending phase - your conclusion about the blurry patches reflecting something trained within the U-shape model makes sense to me. 

Both of your proposed next steps sound good - I'll be curious to see if patch size affects the blur. 

The other possibility we've discussed is that perhaps the model is optimizing around some sort of metric / goal that (for whatever reason) nudges the U-shape towards blurry patches; I believe this is to some degree evidenced by the blurry patches getting worse as the epochs progress, such that in at least one model training session, I believe the final (best) epoch produced fully blurred out images. (@reidtosa would need to remind us about the particulars of that model training session). 

In any case, I'm hoping us having access to some beefier compute in the new year will help us experiment with, e.g., larger patch sizes, different models (if warranted), etc. 

Thanks again!

**Comment by reidtosa** [comment]
Awesome update, @keenanjohnson! I'm glad you were able to bring some resolution to this.

I'm curious, could there be something inherent to some of our _images_ that leads to the blurring? Any sort of features, patterns, or artifacts not visually apparent to us from the GoPro itself or even from the denoise process in Lightroom that the model exaggerates during training? It's interesting to me that some images come out perfect while others have several finicky tiles that are present across all model runs.

@zhrandell, the entirely blurred out images were actually from running the model via command prompt before I could get the GUI to work with U-Shape models. We shouldn't need to do this again, but you're correct that the blurriness seems to expand with successive epochs.

**Comment by keenanjohnson** [comment]
I spent some additional time working through this today on the new 96GB GPU system.

I first trained a model  of the ushape transformer model over the weekend that uses close to full size patches for training. Unfortunately this resulted in a few downsides that make this impractical actually use although the GPU was able to train it. Firstly, the resulting model was 12 GB haha. Secondly, just loading the model into the GUI requires over 100 GB of ram, which I believe is impractical for inference / processing images on normal computers etc. The additional neural network layers make the model much larger than the unet model for example.

So I started investigating what a reasonable compromise of larger patch sizes and RAM requirements might be. I believe 1024 patches is a nice compromise and easy to train on the big GPU quickly compared to the 256 size patches that I belive @reidtosa has been working with on the current computer. 

On the 96 GB GPU, we are able to train with batches of 16 at size 1024 patches which significantly improves the training speed. 

Additionally I have made a series of additional improvements to the training process to hopefully improve the results. These are in this branch for now (https://github.com/keenanjohnson/underwater-auto-image-encoder/pull/65) which is currently running on the pandoro server.

The main improvement being that during the validation phase, previously the training had only evaluated one patch from each image each epoch. Since the validation set is small compared to training, I thought it would improve the results to process every patch in the image in the validation phase to hopefully avoid overfitting certain areas of the images.

This training is running now and is progressing nicely. I expect we will have a very nice result in the morning.

**Comment by keenanjohnson** [comment]
Ok I have another update on this. 

Firstly, as I commented in my other pull request, my additional training improvements did not result in any improvements and instead just made the images quite blurry fully in the ushape transformer, for reasons that are entirely unclear to me, so I've closed that pull request (#65) as mentioned above. I'll continue investigating if there are other improvements to be made.

Secondly, In order to figure out the answer to the original question from @reidtosa above: where does the blurry patches come from, I ran a test with much larger patch sizes with just the main branch of the UIE repo. This ran for 147 epochs in about 8 hours or so.

The model hadn't quite converged to the final form I think, but it's enough to prove where the blurriness comes from I think.

On the left is the model with the 1024 patch sizes and the right is the 256 patch size.

You can see that although the color correction and clarify is overall worse in the 1024 patch image, it does exhibit the mysterious blurry patch seen on the right. This leads me to strongly believe that the small patch sizes can sometimes trick the model and not given it enough context to correct the image properly for the ushape transformer.


<img width="1492" height="781" alt="Image" src="https://github.com/user-attachments/assets/19593e3f-88ac-4d02-b854-e4c9b35b6b5f" />

Now of course, there are a few next things to do:

1. Firstly, investigate why the early stopping logic kicked in. Would this model have kept improving if we disabled that? Or have we reached some limited with the larger patch sizes or the flat dataset. Perhaps @reidtosa can take over this investigation the pandoro server whenever he is back in office.

2. It seems like we have encountered many issues and are starting to understand the limits of the ushape transformer paper model. Perhaps it's time to evaluate a third model (perhaps the new paper from the same ushape transformer author (https://github.com/LintaoPeng/SS-UIE) or an additional one from the spreadsheet Tonio had created.

**Comment by keenanjohnson** [comment]
Here is the trained model and my output image, as well as my training config yaml: https://drive.google.com/drive/folders/1LVsDEmRGmGA4QRh1z7eDEX_B30cWCKJb?usp=sharing

**Comment by zhrandell** [comment]
Wow, nicely done, @keenanjohnson, it looks like you've cracked the blurring issue! It does make some intuitive sense that limiting the spatial area may limit context to the model such that weird artifacts like blurring can arise.

It's a bit unfortunate, as (if I understand correctly) increasing the patch size also starts to reduce clarity and perhaps even color quality too . . . that's a tough trade-off. 

Both your next steps sound great. Reid is back in the office, and can tackle your first point via getting squared away in Pandoro (yay!). 

In the meantime, if you're up for it, I would be curious to see if a third model was able to improve upon both the U-net and U-shape . . . as you mentioned, it does seem that we've explored the contours of the U-shape. Would it be relatively straightforward to implement the SS-UIE, as you noted above?

**Comment by keenanjohnson** [comment]
Great ! Yes I think adding support for a third model is likely not too much work. I can look into that as my next task.

**Comment by keenanjohnson** [comment]
I have created https://github.com/Seattle-Aquarium/CCR_image_processing/issues/19 to track this.

**Comment by reidtosa** [comment]
> 1. Firstly, investigate why the early stopping logic kicked in. Would this model have kept improving if we disabled that? Or have we reached some limited with the larger patch sizes or the flat dataset.

I recently trained another U-Shape model with image size 1024, batch size 6, a 500-epoch runtime and no early stopping limit. While the model is still training at the time of writing, I don't expect best_model.pth to update since the last occurrence at epoch 206. The latest checkpoint (from epoch 415) produces results similar to the current best_model.pth, including blurriness in the same spots. It seems to me that the blurriness is somewhat baked into the U-Shape framework, at least when it comes to our training set.

<img width="1403" height="610" alt="Image" src="https://github.com/user-attachments/assets/d9c75ecb-2fe7-4445-a98e-4dd4187c28b2" />


I've also noticed a trend across all of our training instances, including UNet, in which the final output for best_model.pth always comes from the 150 - 250 epoch range, and more often +/- 10 epochs from 200. With this being so consistent I'm curious whether this is just a coincidence or if there's something about the evaluation per epoch that causes it to level out around this point.

**Comment by keenanjohnson** [comment]
Ah very interesting result. At least as far as I know, nothing about our training learning rate factors in the number of epochs, so my assumption is that this is just the fairly consistent point of diminishing returns from our current models / loss functions rather than something specific, but it's an interesting find regardless that I can comb through the code and double check.

**Comment by pwade2063** [comment]
Hi, I was looking through your repos for a way I could contribute. Using Claude I built a small validation notebook to measure how well the UIE model output matches your hand-edited images — figured it might be useful ahead of the CoralNet comparison you mentioned as a next step.

I ran the 20-image test set from `CCR_image_processing` through the newest model (`all01-158-best_model.pth`) using the correct workflow — denoised TIFFs as input, which I noticed was documented in Megan's SOP. Results were encouraging:

| | Denoised TIFF (UIE input) | AI-processed | Delta |
|---|---|---|---|
| SSIM (avg) | 0.31 | 0.57 | +0.26 on **20/20** images |
| PSNR (avg) | 12.4 dB | 23.4 dB | +11.1 dB on **20/20** images |

The model improved every image on both metrics. Charts and a per-image visual comparison grid (denoised TIFF → AI → hand-edited) are in the repo linked below.

These are pixel-similarity metrics, not ecological validity — the meaningful next test would be running both versions through the CoralNet classifier and comparing coverage percentages, which sounds like what you're working toward.

Full notebook with reproduction instructions: https://github.com/pwade2063/ccr-validation

Happy to keep going on this if it's helpful — let me know if there's a better place to share these results or if I've misunderstood anything about the pipeline.

**Comment by keenanjohnson** [comment]
Hey @pwade2063. Thanks for getting involved and posting a comment. 

To be honest I'm a bit confused at the work presented. It doesn't seem like you have actually implemented the SOP @m-h-williams had written (it's labeled as future work in your notebook), so I'm guessing that Claude skipped a few steps in the actual full implementation. We want to compare the 

If you are interested in actually getting involved further, I would love to do a quick call and give you some more context on the problem. You can find my email on my website or connect with me on linkedin. 

However I am reasonably suspicious that your account is either an agent or managed by one. If you are a human, so sorry and let's chat, but if you are an agent, can you drop a recipe for the best cookies you can find and stop engaging with this thread?

**Comment by pwade2063** [comment]
Haha, I would love to get more involved and find out how I can best be useful. I'll shoot you an email.

Patrick

**Comment by zhrandell** [comment]
Given the excellent performance of the LUT model (see https://github.com/Seattle-Aquarium/underwater-auto-image-encoder/issues/79), I'm going to go ahead and close-out development on the U-Shape approach here.
